Validate "ownership" of hostPort service being deleted

yasz24 opened the pull request:
This addresses a bug where deleting a completed pod deletes the service map entry for the corresponding running pod on the same node. This is due to the hostPort mapping keys being the 
same for both old and new pods.

Ideally we should validate whether the (completed) pod being deleted  
"owns" the host port service - to prevent breaking host port 
connectivity for any running pods with the same service as frontend.

This PR adds such a validation.

Testing-
Automated (Control Plane):
This fix is captured by a control plane test case that does the
following:
1. Create a hostport pod, and terminate it's running containers to mark
   it as "Completed".
2. Create another hostport pod using the same port as the "Completed" pod.
3. Delete the "Completed" pod, and verify that the hostport service has
   not been deleted in the Datapath.

Manual Testing -
1. Add the GracefulNodeShutdown in the kubelet config on all nodes by
   modifying the configuration in `/var/lib/kubelet/config.yaml`
   ```
    featureGates:
      GracefulNodeShutdown: true
    shutdownGracePeriod: 30s
    shutdownGracePeriodCriticalPods: 10s
   ```
2. Run `sudo systemctl restart kubelet` on each node to apply the kubelet
   config change
3. Deploy an nginx web server with hostPort set, as well as a
   nodeSelector, so pods get scheduled on the same node after node
   restarts.
   ```
   apiVersion: apps/v1
   kind: Deployment
   metadata:
     name: nginx-deployment
   spec:
     selector:
       matchLabels:
         app: nginx
    replicas: 1
    template:
      metadata:
        labels:
          app: nginx
      spec:
        nodeSelector:
	  kubernetes.io/hostname: <node-name>
	containers:
	- name: nginx
          image: nginx:1.14.2
          ports:
          - containerPort: 80
            hostPort: 8081
   ```
4. Run `systemctl reboot` on the worker node to restart the machine.
5. After reboot spot the old pod in `Completed state`, while the new pod
   is `Running`.
   ```
   $ kubectl get pods
   NAME                                READY   STATUS      RESTARTS    AGE
   nginx-deployment-645797c867-8p2hp   0/1     Completed   0           13m
   nginx-deployment-645797c867-dx2m8   1/1     Running     0           4m2s
   ```
6. `curl nodeIP:hostPort` successfully get the result.
7. Manually deleted the old pod which is in Completed state.
   ```
   $ kubectl delete pod/nginx-deployment-645797c867-8p2hp
   ```
8. Redo the `curl nodeIP:hostPort`, and successfully get the result again.    // hostPort service has been preserved.

Fixes: #22460

Signed-off-by: Yash Shetty [yashshetty@google.com](mailto:yashshetty@google.com)

aanm reviewed (changes requested):
Thank you for the PR. What happens in case there is only one pod running on the node and that pod is removed? IF we are skipping the generation of service mappings will the entries in the bpf maps be deleted?

aditighag commented on pkg/k8s/watchers/pod.go:
```diff
@@ -503,6 +503,9 @@ func (k *K8sWatcher) genServiceMappings(pod *slim_corev1.Pod, podIPs []string, l
 	)
+	if !k8sUtils.IsPodRunning(pod.Status) {
+		return svcs
+	}
```
This logic is common to add/update/delete watcher callbacks, so this will leak an entry (<host-ip, host-port>) on a cluster. See http://github.com/cilium/cilium/blob/master/pkg/k8s/watchers/pod.go#L271-L271.

@aanm Why do we need to [call](http://github.com/cilium/cilium/blob/master/pkg/k8s/watchers/pod.go#L220-L220) `deleteK8sPodV1` for non-running pods from the **add** watcher callback? We need to clean up previously saved state only when there is an update for pod state, no?

<img width="421" alt="Screen Shot 2022-12-06 at 4 03 16 PM" src="https://user-images.githubusercontent.com/20311845/206052413-d3de500a-b071-4323-996f-1133703c83c8.png">

yasz24 replied:
I see your point about the leaks. The problem essentially boils down to distinguishing between hostPort pods in `completed` status that have a corresponding `running` status pod, versus `completed` pods that don't have a `running` pod.  I think one way to accomplish this could be to add a guard clause when DeleteService is called ([here](https://github.com/cilium/cilium/blob/master/pkg/k8s/watchers/pod.go#L722), and maybe [here](https://github.com/cilium/cilium/blob/master/pkg/k8s/watchers/pod.go#L665) too?), to check the backends IPs for the service being deleted, and validate that it does in fact contain the IP for the completed pod as one of the backends. That way we know that the service being deleted "belongs" to the `completed` pod, and we won't mistakenly break hostPort connectivity for the corresponding `running` pod on the same node, by deleting the service whose "ownership" has since transferred to the `running` pod. The question is whether there is an easy way to get the service's map entries (current backends) in `pkg/k8s/watchers/pod.go` to perform this check? I'm not super familiar with this code, so any thoughts would be appreciated!

aditighag replied:
>The problem essentially boils down to distinguishing between hostPort pods in completed status that have a corresponding running status pod, versus completed pods that don't have a running.

Given that k8s allows only 1 unique <host ip, host port> pod to be deployed at a time, don't we expect that the cilium agent will first receive an event for the completed pod followed by an event for the running pod?

aanm replied:
> > The problem essentially boils down to distinguishing between hostPort pods in completed status that have a corresponding running status pod, versus completed pods that don't have a running.
> 
> Given that k8s allows only 1 unique <host ip, host port> pod to be deployed at a time, don't we expect that the cilium agent will first receive an event for the completed pod followed by an event for the running pod?

@aditighag no, they can be random.

aditighag reviewed (changes requested):
👋 The fix doesn't look right to me, see the inline comment.

yasz24 committed:
Validate "ownership" of hostPort service being deleted

This addresses a bug where deleting a completed pod would wind up
deleting the service map entry for the corresponding running pod on
the same node, due to the hostPort mapping key being the same for
the old and the new pod.

Ideally we want to validate whether the completed pod "owns" the
host port service, before deleting it, thus preventing breakage of
host port connectivity for any running pods with the same service
as frontend.

This commit adds such a validation.

Testing-
Automated (Control Plane):
This fix is captured by a control plane test case that does the
following:
1. Create a hostport pod, and terminate it's running containers to mark
   it as "Completed".
2. Create another hostport pod using the same port as the "Completed" pod.
3. Delete the "Completed" pod, and verify that the hostport service has
   not been deleted in the Datapath.

Manual Testing -
1. Add the GracefulNodeShutdown in the kubelet config on all nodes by
   modifying the configuration in `/var/lib/kubelet/config.yaml`
   ```
    featureGates:
      GracefulNodeShutdown: true
    shutdownGracePeriod: 30s
    shutdownGracePeriodCriticalPods: 10s
   ```
2. Run `sudo systemctl restart kubelet` on each node to apply the kubelet
   config change
3. Deploy an nginx web server with hostPort set, as well as a
   nodeSelector, so pods get scheduled on the same node after node
   restarts.
   ```
   apiVersion: apps/v1
   kind: Deployment
   metadata:
     name: nginx-deployment
   spec:
     selector:
       matchLabels:
         app: nginx
    replicas: 1
    template:
      metadata:
        labels:
          app: nginx
      spec:
        nodeSelector:
	  kubernetes.io/hostname: <node-name>
	containers:
	- name: nginx
          image: nginx:1.14.2
          ports:
          - containerPort: 80
            hostPort: 8081
   ```
4. Run `systemctl reboot` on the worker node to restart the machine.
5. After reboot spot the old pod in `Completed state`, while the new pod
   is `Running`.
   ```
   $ kubectl get pods
   NAME                                READY   STATUS      RESTARTS    AGE
   nginx-deployment-645797c867-8p2hp   0/1     Completed   0           13m
   nginx-deployment-645797c867-dx2m8   1/1     Running     0           4m2s
   ```
6. `curl nodeIP:hostPort` successfully get the result.
7. Manually deleted the old pod which is in Completed state.
   ```
   $ kubectl delete pod/nginx-deployment-645797c867-8p2hp
   ```
8. Redo the `curl nodeIP:hostPort`, and successfully get the result again.    // hostPort service has been preserved.

Signed-off-by: Yash Shetty <yashshetty@google.com>

squeed commented on pkg/k8s/watchers/pod.go:
```diff
@@ -719,6 +719,11 @@ func (k *K8sWatcher) deleteHostPortMapping(pod *slim_corev1.Pod, podIPs []string
 	for _, dpSvc := range svcs {
+		svcs := k.svcManager.GetDeepCopyServices()
+		svc := findServiceByL3n4Addr(dpSvc.Frontend.L3n4Addr, svcs)
+		if svc != nil && !deepEqualBackends(svc.Backends, dpSvc.Backends) {
```
makes sense; can you add a comment for what this does?

yasz24 replied:
will do

tommyp1ckles commented on pkg/k8s/watchers/pod.go:
```diff
@@ -719,6 +719,11 @@ func (k *K8sWatcher) deleteHostPortMapping(pod *slim_corev1.Pod, podIPs []string
 	}
 
 	for _, dpSvc := range svcs {
+		svcs := k.svcManager.GetDeepCopyServices()
```
Copying all of these, in each loop iteration, seems like it might be a bit heavy?

yasz24 replied:
Good catch! Yes, there is definitely no need for the service copies to be retrieved in each iteration, and the operation can definitely be moved outside the for-loop. Will go ahead and do so 👍

squeed replied:
commented below; we can simplify this even more :-)

yasz24 replied:
done

squeed commented on pkg/k8s/watchers/pod.go:
```diff
@@ -719,6 +719,11 @@ func (k *K8sWatcher) deleteHostPortMapping(pod *slim_corev1.Pod, podIPs []string
 
 	for _, dpSvc := range svcs {
+		svcs := k.svcManager.GetDeepCopyServices()
+		svc := findServiceByL3n4Addr(dpSvc.Frontend.L3n4Addr, svcs)
```
You can skip all this looping (and the full deepcopy above) by adding this method to the ServiceManager:

```go
func (s *Service) GetDeepCopyServiceByFrontend(frontend lb.L3n4Addr) (*lb.SVC, bool) {
	s.Lock()
	defer s.Unlock()

	if svc, found := s.svcByHash[frontend.Hash()]; found {
		return svc.deepCopyToLBSVC(), true
	}

	return nil, false
}
```

Then you won't need `findServiceByL3nL4Addr` or the full `GetDeepCopyServices`. Saves a lot of allocs and looping.

yasz24 replied:
Done!

aditighag reviewed (commented):
Logic looks good to me. Please incorporate Casey's suggestion in the implementation. Thanks! How did you test the fix, and I was wondering if you had any thoughts on adding tests? See - https://github.com/cilium/cilium/tree/master/test/controlplane.

aditighag commented on pkg/k8s/watchers/pod.go:
```diff
@@ -719,6 +719,11 @@ func (k *K8sWatcher) deleteHostPortMapping(pod *slim_corev1.Pod, podIPs []string
 	}
 
 	for _, dpSvc := range svcs {
+		svcs := k.svcManager.GetDeepCopyServices()
```
The ownership check also warrant a code comment explaining the motivation.

yasz24 replied:
Done!

tommyp1ckles commented on pkg/k8s/watchers/pod.go:
```diff
@@ -728,6 +733,28 @@ func (k *K8sWatcher) deleteHostPortMapping(pod *slim_corev1.Pod, podIPs []string
+	if len(backends1) != len(backends2) {
+		return false
+	}
+	for idx, backend1 := range backends1 {
```
are we sure these will always be in the same order?

yasz24 replied:
Good point, i'm not sure if we can assume they will be in the same order. Updated the deepEquals logic to be order agnostic. PTAL

squeed commented on pkg/k8s/watchers/pod.go:
```diff
@@ -731,6 +742,37 @@ func (k *K8sWatcher) deleteHostPortMapping(pod *slim_corev1.Pod, podIPs []string
 	return nil
 }
 
+func findServiceByL3n4Addr(l3n4Addr loadbalancer.L3n4Addr, svcs []*loadbalancer.SVC) *loadbalancer.SVC {
```
You can delete this function now :-)

yasz24 replied:
yes :)

squeed commented on pkg/k8s/watchers/watcher.go:
```diff
@@ -141,6 +141,8 @@ type policyRepository interface {
 
 type svcManager interface {
 	DeleteService(frontend loadbalancer.L3n4Addr) (bool, error)
+	GetDeepCopyServices() []*loadbalancer.SVC
```
We can remove this from the interface now, right?

yasz24 replied:
removed!

squeed commented on pkg/k8s/watchers/pod.go:
```diff
@@ -731,6 +742,37 @@ func (k *K8sWatcher) deleteHostPortMapping(pod *slim_corev1.Pod, podIPs []string
+
+	for _, backend2 := range backends2 {
+		backend1 := l3n4AddrMap[backend2.L3n4Addr.AddrCluster.String()]
+		if backend1.L3n4Addr.DeepEqual(&backend2.L3n4Addr) {
```
need to add a nil check here too

squeed replied:
(backend1 could be nil)

yasz24 replied:
Added!

aditighag commented on pkg/k8s/watchers/pod.go:
```diff
@@ -731,6 +742,37 @@ func (k *K8sWatcher) deleteHostPortMapping(pod *slim_corev1.Pod, podIPs []string
+		return false
+	}
+
+	l3n4AddrMap := make(map[string]*loadbalancer.Backend)
```
The slice length is known, so might as well use it?

Edit: Didn't realize that you were creating a map, nvm.

yasz24 replied:
Not sure which slice you're referring to?

aditighag commented on pkg/k8s/watchers/pod.go:
```diff
@@ -731,6 +742,37 @@ func (k *K8sWatcher) deleteHostPortMapping(pod *slim_corev1.Pod, podIPs []string
+	}
+
+	for _, backend2 := range backends2 {
+		backend1 := l3n4AddrMap[backend2.L3n4Addr.AddrCluster.String()]
```
```suggestion
		if backend1, ok := l3n4AddrMap[backend2.L3n4Addr.AddrCluster.String()]; ok {
		:
		}
```

yasz24 replied:
Done!

aditighag commented on pkg/k8s/watchers/pod.go:
```diff
@@ -731,6 +742,27 @@ func (k *K8sWatcher) deleteHostPortMapping(pod *slim_corev1.Pod, podIPs []string
+	}
+
+	for _, backend2 := range backends2 {
+		if backend1, ok := l3n4AddrMap[backend2.L3n4Addr.AddrCluster.String()]; ok && backend1.L3n4Addr.DeepEqual(&backend2.L3n4Addr) {
```
Why call the `DeepEqual` method if there was a look up done based on the `L3nL4Addr` string?

yasz24 replied:
Good point. Could the `L4Addr`  or `Scope` in the `L3n4Addr` differ for two backends with the same IP? If not, i think we could definitely get rid of the `DeepEqual` check 👍

yasz24 replied:
Actually, now that I think of it, I think we could just use the `backend2.String()` as the map key, instead of using the `backend2.L3n4Addr.AddrCluster.String()`. Then we don't need the `DeepEqual` check. Will update 👍

yasz24 replied:
Done!

aditighag commented on pkg/k8s/watchers/pod.go:
```diff
@@ -733,6 +744,27 @@ func (k *K8sWatcher) deleteHostPortMapping(pod *slim_corev1.Pod, podIPs []string
+	l3n4AddrMap := make(map[string]*loadbalancer.Backend)
+
+	for _, backend1 := range backends1 {
+		l3n4AddrMap[backend1.String()] = backend1
```
Why not use `backend1.L3n4Addr.String()` for comparing backends? The backend structure has different attributes besides the L3nL4 address, and while the `String()` method currently returns only the L3nL4 address in the string format, it's better to not rely on that being the case (especially since we don't have any tests covering this logic).

yasz24 replied:
Makes sense. Agreed that it's probably better to be explicit. Updated!

aanm commented on pkg/service/service.go:
```diff
@@ -944,6 +944,18 @@ func (s *Service) GetDeepCopyServicesByName(name, namespace string) (svcs []*lb.
+// GetDeepCopyServiceByFrontend returns a deep-copy of the service that matches the Frontend address.
+func (s *Service) GetDeepCopyServiceByFrontend(frontend lb.L3n4Addr) (*lb.SVC, bool) {
+	s.Lock()
+	defer s.Unlock()
```
```suggestion
	s.RLock()
	defer s.RUnlock()
```

aanm reviewed (changes requested):
Just some small nits from my side. @aditighag where can we add some tests covering this?

aanm commented on pkg/k8s/watchers/pod.go:
```diff
@@ -736,6 +747,27 @@ func (k *K8sWatcher) deleteHostPortMapping(pod *slim_corev1.Pod, podIPs []string
+	}
+
+	for _, backend2 := range backends2 {
+		if _, ok := l3n4AddrMap[backend2.L3n4Addr.String()]; ok {
```
```suggestion
	l3n4AddrMap := make(map[loadbalancer.L3n4Addr]struct{})

	for _, backend1 := range backends1 {
		l3n4AddrMap[backend1.L3n4Addr] = struct{}{}
	}

	for _, backend2 := range backends2 {
		if _, ok := l3n4AddrMap[backend2.L3n4Addr]; ok {
```

aanm reviewed (changes requested):
This should have been part of the previous review

aojea commented on test/controlplane/pod/hostport/hostport.go:
```diff
@@ -0,0 +1,58 @@
+	test.
+		UpdateObjectsFromFile(abs("init.yaml")).
+		SetupEnvironment(modConfig).
+		StartAgent().
```
@joamaki we are stuck with the test, it hangs at this point and it never move forward, we compared with the other tests and we can't see anything obvious, can you help us to understand what is missing?

It seems that it is waiting for a local node initialization
```
oroutine 118 [chan receive, 9 minutes]:
github.com/cilium/cilium/daemon/cmd.(*localNodeInitializer).getK8sLocalNode(0xc000e9c6f0, {0x3ed7ef8?, 0xc000527b80?})
        /usr/local/google/home/aojea/src/cilium/daemon/cmd/local_node_init.go:87 +0xec
github.com/cilium/cilium/daemon/cmd.(*localNodeInitializer).initFromK8s(0xc000e9c6f0, {0x3ed7ef8?, 0xc000527b80?}, 0xc0001ea178)
        /usr/local/google/home/aojea/src/cilium/daemon/cmd/local_node_init.go:101 +0x5b
github.com/cilium/cilium/daemon/cmd.(*localNodeInitializer).InitLocalNode(0x4b1886?, {0x3ed7ef8, 0xc000527b80}, 0xc0001ea178)
        /usr/local/google/home/aojea/src/cilium/daemon/cmd/local_node_init.go:48 +0x9e
github.com/cilium/cilium/pkg/node.NewLocalNodeStore.func1({0x7f30000e7d00?, 0xc000527b80})
        /usr/local/google/home/aojea/src/cilium/pkg/node/local_node_store.go:81 +0x187
github.com/cilium/cilium/pkg/hive.Hook.Start(...)
        /usr/local/google/home/aojea/src/cilium/pkg/hive/lifecycle.go:35
github.com/cilium/cilium/pkg/hive.(*DefaultLifecycle).Start(0xc00123e420, {0x3ed7f30?, 0xc000226008?})
        /usr/local/google/home/aojea/src/cilium/pkg/hive/lifecycle.go:104 +0x2ee
github.com/cilium/cilium/pkg/hive.(*Hive).Start(0xc00127cb40, {0x3ed7f30, 0xc000226008})
        /usr/local/google/home/aojea/src/cilium/pkg/hive/hive.go:295 +0x105
github.com/cilium/cilium/test/controlplane/suite.startCiliumAgent(0xc000303ba0, {0x3f17488?, 0xc0012291c0}, {0x3ec03d0?, 0x63b9620})
        /usr/local/google/home/aojea/src/cilium/test/controlplane/suite/agent.go:105 +0x785
github.com/cilium/cilium/test/controlplane/suite.(*ControlPlaneTest).StartAgent(0xc001261770, {0x0, 0x0, 0x0})
        /usr/local/google/home/aojea/src/cilium/test/controlplane/suite/testcase.go:141 +0xe5
github.com/cilium/cilium/test/controlplane/services/hostport.testHostPort(0xc000303ba0)
        /usr/local/google/home/aojea/src/cilium/test/controlplane/services/hostport/hostport.go:42 +0x1f2
testing.tRunner(0xc000303ba0, 0x3ae3388)
        /usr/local/google/home/aojea/sdk/go1.20.4/src/testing/testing.go:1576 +0x10b
created by testing.(*T).Run
        /usr/local/google/home/aojea/sdk/go1.20.4/src/testing/testing.go:1629 +0x3ea
```

aojea replied:
found the problem https://github.com/cilium/cilium/pull/22587/files#r1244016712

aojea commented on test/controlplane/pod/hostport/hostport.go:
```diff
@@ -0,0 +1,58 @@
+
+	k8sVersions := controlplane.K8sVersions()
+	// We only need to test the last k8s version
+	test := suite.NewControlPlaneTest(t, "pod-hostport-control-plane", k8sVersions[len(k8sVersions)-1])
```
```suggestion
	test := suite.NewControlPlaneTest(t, "hostport-control-plane", k8sVersions[len(k8sVersions)-1])
```

this has to match teh name of the control plane node on the init.yaml file, otherwise it does not start waiting for the local node

yasz24 replied:
done!

aojea commented on test/controlplane/pod/hostport/generate.sh:
```diff
@@ -0,0 +1,60 @@
+cilium install --wait
+
+: Dump the initial state
+kubectl get nodes,pods -o yaml > "${dir}/init.yaml"
```
```suggestion
kubectl get nodes,ciliumnodes,pods -o yaml > "${dir}/init.yaml"
```

with your PR it fails in one endpoint to me, I regenerated and also added this , but I'm not sure this makes a difference

yasz24 replied:
I don't think this is necessary (getting ciliumnodes). Passes for me as is.

aojea commented on test/controlplane/pod/hostport/manifests/hostport-2.yaml:
```diff
@@ -0,0 +1,19 @@
+      periodSeconds: 20
+      httpGet:
+        path: /
+        port: 80
```
end of line

yasz24 replied:
done

aojea commented on test/controlplane/pod/hostport/manifests/hostport-1.yaml:
```diff
@@ -0,0 +1,19 @@
+      periodSeconds: 20
+      httpGet:
+        path: /
+        port: 80
```
add end of line

yasz24 replied:
done

squeed reviewed (approved):
nice work!

nebril commented on pkg/k8s/watchers/pod.go:
```diff
@@ -782,6 +793,27 @@ func (k *K8sWatcher) deleteHostPortMapping(pod *slim_corev1.Pod, podIPs []string
 	return nil
 }
 
+func deepEqualBackends(backends1 []*loadbalancer.Backend, backends2 []*loadbalancer.Backend) bool {
```
Would be nice to get a unit test for this function.

yasz24 replied:
Done!

nebril commented on pkg/k8s/watchers/pod.go:
```diff
@@ -782,6 +793,27 @@ func (k *K8sWatcher) deleteHostPortMapping(pod *slim_corev1.Pod, podIPs []string
 	return nil
 }
 
+func deepEqualBackends(backends1 []*loadbalancer.Backend, backends2 []*loadbalancer.Backend) bool {
```
AFAIU this function doesn't care about backend order in provided slices, can you add a comment about that?

yasz24 replied:
Done!

nebril reviewed (commented):
Thanks for the PR! LGTM overall, please add the missing unit test, otherwise it's good to go!
